ABC Simulator

kspieks opened the pull request:
This PR creates a template for adding SA adapters. It implements SA with RMG. I welcome all feedback and look forward to your comments :)

kspieks commented:
Hmm the Travis tests failed due to an import error from RMG-Py even though the master branch of RMG-Py passes all tests. I also tried running `coverage run -m pytest -ra`, which is the same command travis uses, and it passed locally for me giving the following output

```
coverage run -m pytest -ra
============================================================ test session starts ============================================================
platform darwin -- Python 3.7.6, pytest-5.4.1, py-1.8.1, pluggy-0.13.1
rootdir: /Users/kevin/Documents/Green_Lab/RMG/T3, inifile: pytest.ini
collected 18 items                                                                                                                          

tests/test_main.py ................                                                                                                   [ 88%]
tests/test_utils.py ..                                                                                                                [100%]

============================================================ 18 passed in 13.31s ============================================================

```

I must be missing something obvious. Can you help me understand why these imports failing?

alongd commented:
T3 uses RMS's binaries from conda...
For now we can import RMG's master for the tests. Let's make a separate PR for that

kspieks commented:
> T3 uses RMS's binaries from conda...
> For now we can import RMG's master for the tests. Let's make a separate PR for that

Sorry what files must we update in a separate PR? I'm not seeing where the RMS binaries are specified.

alongd commented:
I had a typo, it's RMG, not RMS. I'll make the PR

kspieks commented:
> I had a typo, it's RMG, not RMS. I'll make the PR

Okay thanks for making the PR. Admittedly I'm still confused where RMG's conda binaries entered the travis build so I'll learn from your PR

kspieks commented:
@alongd Can you review this PR when you get a chance?

alongd commented on tests/test_rmg_sa.py:
```diff
@@ -0,0 +1,93 @@
+###############################################################################
+
+"""
+This module contains unit tests of the :mod:`tandem.sa.rmg_sa` module.
```
We use a convention of the module "path", see for example https://github.com/ReactionMechanismGenerator/T3/blob/master/tests/test_utils.py

kspieks replied:
Ah I see. So I should use this instead?
```
"""
t3 tests test_rmg_sa module
"""
```

alongd replied:
Yes

alongd commented on tests/test_rmg_sa.py:
```diff
@@ -0,0 +1,93 @@
+        """
+        A method that is run after all unit tests in this class.
+        """
+        log_file = os.path.join(cls.output_directory, 't3.log')
```
maybe create a list of files, then iteratively remove them?
see example: https://github.com/ReactionMechanismGenerator/ARC/blob/master/arc/restartTest.py#L205

kspieks commented on tandem/sa/factory.py:
```diff
@@ -0,0 +1,96 @@
+    _registered_sa_adapters[sa] = sa_class
+
+
+def sa_factory(sa_method: str,  # add everything that goes into the adapter class init
```
It was just a note for future developers, but I'll remove it as it's not necessary

kspieks commented on tandem/sa/adapter.py:
```diff
@@ -0,0 +1,48 @@
+# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER         #
+# DEALINGS IN THE SOFTWARE.                                                   #
+#                                                                             #
+###############################################################################
```
From the discussion on [RMG-Py PR#1929](https://github.com/ReactionMechanismGenerator/RMG-Py/pull/1929), should we include a copywrite header when removing the license such as:

```
# Copyright (c) 2002-2020 Prof. William H. Green (whgreen@mit.edu),           
# Prof. Richard H. West (r.west@neu.edu), and the RMG Team (rmg_dev@mit.edu). 
# Please see the LICENSE.txt in the root directory for more information.
```

alongd replied:
In this repo we don't need a copyright comment on each file

kspieks replied:
Sounds good. I'll remove the copyright from each file in this PR

kspieks commented on tandem/sa/factory.py:
```diff
@@ -0,0 +1,96 @@
+    if sa_method.lower() not in ['rmg', 'rms', 'cantera']:
+        raise InputError(f'The "SA method" argument must equal "RMG", "RMS", or "Cantera". Got: {sa_method}'
+                         f'Currently only RMG is implemented as an SA method')  # temp
+    print(f'ln 87 in factory. sa_method is: {sa_method}')
```
Grr I thought I removed all of those. Wasn't careful enough. Thanks!

kspieks commented on tandem/sa/rmg_sa.py:
```diff
@@ -0,0 +1,165 @@
+        Simulate a job, run a sensitivity analysis, and generate output csv files using RMG.
+
+        Raises:
+            InputError: If a given observable in ``observable_list`` does not have structure (smiles or adjacency list).
```
I assume I should replace both `InputError`s with `ValueError`s?

alongd replied:
Yes, let's not use RMG imported Errors (other than for catching RMG errors)

alongd commented:
Let's asses what parameters RMS/Cantera require, and make sure we have them in out SAAdapter class. I think the most complicated thing is the reactor type: for RMG we simply use the reactor type from the input file, but here we should let the user specify how to simulate and run SA.

xiaoruiDong commented:
Thank you for the PR. I totally support the transition to adapters Since sensitivity is also a simulation just with more things to calculate, I don't know if you agree that we should have a simulation class (adapter) first and build SA adapter on that.

Below is a link a cantera sensitivity input example:
https://projects.skill-lync.com/projects/Sensitivity-Analysis-of-GRI30-Mechanism-using-Cantera-36940
To me, the simulation input is just like a subset to the sensitivity input.

Besides, I saw there is an argument (input_file), do we want to pursue a workflow of writing the input file first and then apply it to this SA factory, or it is just the input_file of RMG we used to generate the model?

If the first case, the observable list can be removed (included in the input file). I think it introduces more flexibility.  For example, I use range reactor in the model generation, while using a range of different conditions to do sensitivities. Besides, adiabatic conditions are also something I considered for sensitivity. Besides, in combustion community, sensitivity analysis on ignition delay is another thing wanted. This is probably also something we want to add into consideration. So, since we want to build input_files, we need to have a template write. (Jinja2 is something we want to pursue, I guess. Its inheritance function is useful).

If the second case, I feel that we will be limited, but probably let me know your reason first.

alongd commented:
good point @xiaoruiDong, let's make a SimulateAdapter instead, we could have a bool flag for requesting SA in addition.

Regarding the other points, we're just starting to discuss how to use Cantera and RMS, this part isn't ripe yet. Yes, it's possible to ask the user to supply an additional script for the simulation, specifying the reactors.

kspieks commented:
Thank you both for this valuable feedback! I agree that the first case you describe offers better generalizability. I'll start working on a new branch called `abc simulate` that tries to incorporate these ideas: creates a simulate class that uses an input file separate from the RMG input.py and ARC arc.yml files passed to T3

I have some clarifying questions for you both:

1. So the big idea is that once T3 has completed model generation, we allow the user to simulate the model with either RMG, RMS, or Cantera? As an additional feature, the user can also specify which software to use when performing SA?

2. What motivates performing SA for different T, P, or in a different reactor type than what was used to generate the model? I didn't know this was a common need, which is why I hadn't planned for it when writing the SA adapter for this PR.

kspieks commented:
I was in the process of addressing Alon's comments so I just pushed the changes for all resolved conversations. Ultimately, I will start working on a new branch

alongd commented:
I think we don't necessarily need a new branch.
In the grand scheme of things, T3 will also be able to generate a report with fluxes, profiles, etc.:
![image](https://user-images.githubusercontent.com/16158262/80033468-cb6e2c00-84ba-11ea-89de-7d5317c9f396.png)

Since RMG has limited reactor types (which is OK for model generation), the process being modeled might be different and require a more complicated reactor model.

kspieks commented:
> I think we don't necessarily need a new branch.
> In the grand scheme of things, T3 will also be able to generate a report with fluxes, profiles, etc.:
> ![image](https://user-images.githubusercontent.com/16158262/80033468-cb6e2c00-84ba-11ea-89de-7d5317c9f396.png)
> 
> Since RMG has limited reactor types (which is OK for model generation), the process being modeled might be different and require a more complicated reactor model.

I see. I was unaware that developing a model in reactor A and then simulating in reactor B gave good results. Good to know!

Okay I can keep this branch. I was considering creating a new Simulate class that has a method for performing SA. But I guess is it better organization to have both Simulate and SA be separate classes?

alongd commented:
I agree with the first approach. We could have one (ABC) class, say SimulateAdaptor, that has many methods such as SA, UA, species profiles, fluxes... and at the core of all these features (I think) is simulating the reactor(s).

kspieks commented:
> I agree with the first approach. We could have one (ABC) class, say SimulateAdaptor, that has many methods such as SA, UA, species profiles, fluxes... and at the core of all these features (I think) is simulating the reactor(s).

I see. Thanks for clarifying. So would simulate just be an additional method that SA, UA, etc would call?

As another question, how much flexibility are we looking for? Would SA, UA, etc actually be abstract methods in an ABC SimulateAdapter? Say we have 3 adapters: RMG, RMS, Cantera. Each has methods for simualte, SA, UA, etc. Normally, we would use the factory to create one instance of the needed adapter. I'm thinking it would be cumbersome for a user to simulate with Cantera and do SA with RMS with this set up. But perhaps this isn't necessary?

kspieks commented:
@alongd Can you take a look at this draft of the `RMGSimulator` when you get a chance?

alongd commented on tandem/simulate/rmg_simulator.py:
```diff
@@ -0,0 +1,311 @@
+        sa_rtol (Optional[float]): The relative tolerance used when running SA.
+        verbose (Optional[bool]): Whether or not to log to file.
+
+    ======================= ======      ====================================================
```
we don't use this style for attributes in T3 (we do in RMG), you can have an `Attributes:` section in a similar format to `Args:`

kspieks replied:
Thanks for reviewing! I just pushed the updated format for documenting the attributes. I think the pyrms simulator adapter is ready as well, just needs SA to be working in rms. I'm happy to help with the T3 input file in the meantime :)

alongd commented on tandem/simulate/rmg_simulator.py:
```diff
@@ -0,0 +1,311 @@
+
+        Returns:
+             A SA dictionary, whose structure is given below.
+             sa_dict = { 'thermo' :
```
excellent, thanks!

alongd commented on t3/simulate/rmg_simulator.py:
```diff
@@ -0,0 +1,273 @@
+                                         check_duplicates=False,
+                                         )
+        if self.rmg_model is None:
+            self.logger.error('The RMG adapter did not properly read the rmg input file.')
```
shall we also return from this call (e.g., we'll crash on `for reaction_system in self.rmg_model.reaction_systems`)? and check this is not None in other method calls?

alongd commented on t3/simulate/rmg_simulator.py:
```diff
@@ -0,0 +1,273 @@
+                                       if species.label in self.observable_list]
+
+            for reaction_system in self.rmg_model.reaction_systems:
+                if isinstance(reaction_system, SimpleReactor):
```
the two reactors have repetivive code, and some unique (P, V). Can we consolidate the repetitive parts?

alongd commented on t3/simulate/rmg_simulator.py:
```diff
@@ -0,0 +1,273 @@
+
+        try:
+            self.logger.info('Simulating a job using RMG...')
+            simulate(self.rmg_model)
```
Can we check whether the file exists (raise an error if it doesn't) instead of the try-except?
Anyway, the `self.logger` statement should be outside the try-except block.

alongd commented on t3/simulate/rmg_simulator.py:
```diff
@@ -0,0 +1,273 @@
+                        }
+        """
+
+        def get_species_by_label(label: str,
```
How is this different than t3.main.get_species_by_label()? Maybe move the original one to common and use it here as well?

alongd commented on t3/simulate/rmg_simulator.py:
```diff
@@ -0,0 +1,273 @@
+        return sa_dict
+
+
+
```
PEP8: please only leave two line breaks between segments

alongd commented on t3/common.py:
```diff
@@ -47,3 +51,34 @@ def delete_root_rmg_log(project_directory: str) -> None:
         os.remove(rmg_log_path)
+
+
+def get_species_by_label(label: str,
```
I see you already did this here, please squash/organize the commits so we won't have addition+deletion in the same PR.

alongd commented on tests/test_rmg_simulator.py:
```diff
@@ -0,0 +1,88 @@
+    t3.iteration = 0
+    t3.set_paths()
+    return t3
+
```
PEP8: have two line breaks between functions

alongd commented on tests/data/rmg_simulator_test/input.yml:
```diff
@@ -0,0 +1,47 @@
+project: T3_minimal
```
Please change the commit message to `Tests: Added RMG Simulator Adapter tests` (see https://github.com/ReactionMechanismGenerator/T3/wiki/Contributor-Guidelines#commits-style)

alongd commented on tests/test_main.py:
```diff
@@ -22,6 +22,7 @@
                      get_species_by_label,
```
Please change the commit message: `Tests: Modifications to... to account for...`

alongd commented on t3/main.py:
```diff
@@ -14,7 +14,6 @@
 
```
The commit message says "Update rmg_simulator.py", the intention is probably "main.py"

kspieks commented:
Thanks for reviewing the updated changes on the RMG adapter! I believe the recently pushed changes incorporate all feedback :)

alongd commented on t3/common.py:
```diff
@@ -47,3 +50,34 @@ def delete_root_rmg_log(project_directory: str) -> None:
+                              Returns ``None`` if no species was found.
+    """
+    if label is None:
+        raise InputError('Got None as label input')
```
Do you mind raising a general error, e.g., `ValueError`? We shouldn't use errors imported from RMG for non-RMG stuff (we have several instances like that on master, let's gradually change them when we catch them)

kspieks replied:
Updated

alongd commented on t3/main.py:
```diff
@@ -128,6 +126,7 @@ def __init__(self,
                  ):
 
         self.t0 = time.time()  # initialize the timer
+        self.sa_dict = None    # initliaze sa_dict
```
We need to think about where to put the sa_dics docstring structure. It's currently in the RMG adapter, perhaps it should be relocated to main? I'm not sure - what do you think?

kspieks replied:
I see what you mean. Let's relocate it to main since every adapter will follow the same structure for it's `get_sa_coefficients()` method and return the same `sa_dict()` to main

alongd commented on t3/main.py:
```diff
@@ -728,46 +677,21 @@ def determine_species_based_on_sa(self) -> List[int]:
                               "Not performing refinement based on sensitivity analysis!")
```
Could you delete the "Adapters notes" in the docstring above?

alongd commented on tests/test_common.py:
```diff
@@ -4,8 +4,33 @@
+from t3.main import T3
+
+
+def run_minimal(project: Optional[str] = None,
```
Should this function be imported from test_main.py? Or should it be relocated to a common file?

kspieks replied:
Yea that's smart. I'll move `run_minimal` to `test_main`

alongd commented on t3/schema.py:
```diff
@@ -495,6 +495,11 @@ def check_species_and_reactors(cls, values):
                                              f'have actual corresponding species.')
         return values
 
+    @validator('options', always=True)
```
👍

alongd commented on tests/test_main.py:
```diff
@@ -147,7 +148,20 @@
                }
-
+rmg_minimal_defaults = rmg_minimal.copy()
+rmg_minimal_defaults['options'] = {'seed_name': 'Seed',
```
Could you explain why we need to copy and add? Can't we add this block to the original `rmg_minimal`?

kspieks replied:
I assumed it was intentional to leave the RMG options block as `None` as you'd defined it so make sure that the schema correctly assigns the defaults. So the updated dictionary `rmg_minimal_defaults` has the values after schema updates the dictionary so that the assert statement is True. Did you want me to change the line of `'options': None,` to reflect my copy and add so it has the arguments from `rmg_minimal_defaults`?

alongd replied:
I see. You're right. No need to add additional modifications.

alongd commented on tests/test_rmg_simulator.py:
```diff
@@ -0,0 +1,102 @@
+TEST_DIR = os.path.join(DATA_BASE_PATH, 'rmg_simulator_test')
+
+
+def run_minimal(project_directory: Optional[str] = None) -> T3:
```
same comment (import function?)

kspieks commented:
True we've changed the structure a bit, and our most recent iteration has the adapter create general SA (or UA later) folder so there is no longer an `rmg_analysis` folder. The RMG adapter uses the input file from the RMG folder to do SA. In the process, it creates the SA folder and copies the input file to there and creates the csvs. Since the SA folder is created during testing, I didn’t think it was necessary to commit the files (like input file and csvs) so I removed them to keep the commit cleaner.

alongd commented:
OK. Didn't we want to re-write the RMG input file in the RMG_simulator adpter?

kspieks commented:
I thought we were only writing the RMG input file in `main.py` and having the RMG adapter simply copy it to the `SA` folder if doing SA. If not performing SA, nothing happens since `input.py` already exists in the `RMG` folder for that iteration so it just simulates the mechanism. What's the anticipated behavior for the adapter?

Sorry to not check for cyclic imports. I pushed the simplest fix, which is to move the `run_minimal` function to a new `common_.py` file to avoid the error.

alongd commented:
OK regarding not re-writing the RMG file, I may have got confused and didn't remember our off-line discussion correctly.

Perhaps open a common.py file under the tests folder instead of having common and common_?

kspieks commented:
No worries. Definitely reach out over slack to discuss further if you want. I moved the file to `tests/common.py`

I also started the RMS adapter. I've been having trouble getting PyCall to work properly and will likely ask for help on getting travis to set up PyCall correctly to run the RMS tests

kspieks commented on tests/test_simulate_adapters/test_cantera_simulator.py:
```diff
@@ -0,0 +1,84 @@
+    """
+    Run RMG's minimal example with SA by testing the `set_up` method within the CanteraSimulator init method.
+    Then run the `get_sa_coefficients()` method to test that Cantera correctly performs SA.
+    to obtain sa_dict.
```
I'll fix this typo when I incorporate addition feedback :)

alongd commented on .travis.yml:
```diff
@@ -20,29 +17,38 @@ before_install:
+      - codecov
+    - script:
+      - source activate t3_env
+      - travis_wait 45 coverage run -m pytest tests/test_simulate_adapters/ -vv
```
nice!
will we get two coverage reports?
why wait 45 (seconds?)?

kspieks replied:
I don't think so? RMG-Py runs multiple tests in [parallel](https://travis-ci.org/github/ReactionMechanismGenerator/RMG-tests/builds/752092205) but only generates one coverage [report](https://github.com/ReactionMechanismGenerator/RMG-Py/runs/1625449533). 

At the time of this PR, SA was slow in pyrms so I made travis wait to avoid timing out. I think Matt and/or Hao-Wei have fixed this bug in RMS since then, but I'd prefer to make any pyrms changes in a future PR because it was very finicky for this PR.

alongd commented on t3/common.py:
```diff
@@ -158,3 +159,33 @@ def convert_termination_time_to_seconds(termination_time: Union[float, str]):
     return t_final
+
+
+def run_ck2cti(chemkin_file: str,
```
RMG _should_ be calling this after model termination. I agree it's a good idea to also allow T3 to have this functionality. Where is this being currently used?

kspieks replied:
RMG does create a cantera file so this helper function is not used by T3. I had added this in case someone wants to use the `CanteraSimulator` for a mechanism not generated by RMG, in which case they may have to convert a chemkin file to a cantera file. Would you prefer I remove this and the corresponding unit test since it's not explicitly used by T3?

alongd replied:
I see. I think the idea is that T3 will generate a model instead of someone feeding it a Chemkin file. If we need ck2cti we could perhaps use [RMG's functionality](https://github.com/ReactionMechanismGenerator/RMG-Py/blob/3c13a22bc02472f8ea577f4bae69d2c381678570/rmgpy/tools/canteramodel.py#L290). I think we shouldn't include it in T3, but I'm not sure. What do you think?

kspieks replied:
That makes sense. I'll drop the 2 commits that add the function and corresponding unit test.

alongd reviewed (commented):
Thanks for this awesome PR!
Sorry for the delay in reviewing
I added some comments, I still haven't reviewed the Cantera adapter and the notebooks.

kspieks commented on t3/simulate/factory.py:
```diff
@@ -0,0 +1,94 @@
+                     atol: float,
+                     rtol: float,
+                     observable_list: Optional[list] = None,
+                     sa_atol: Optional[float] = 1e-6,
```
Thanks this was my mistake. I'll update `sa_atol` and `sa_rtol` to be `float` for the factory and all simulate adapters.

kspieks commented on t3/simulate/rmg_simulator.py:
```diff
@@ -0,0 +1,236 @@
+        Obtain the SA coefficients.
+
+        Returns:
+             sa_dict: a SA dictionary, whose structure is given in the docstring for main.py
```
Thanks for catching this. I added `(dict)` to this method in each of the adapters

kspieks commented on t3/common.py:
```diff
@@ -17,6 +17,7 @@
 DATA_BASE_PATH = os.path.join(t3_path, 'tests', 'data')
 EXAMPLES_BASE_PATH = os.path.join(t3_path, 'examples')
 PROJECTS_BASE_PATH = os.path.join(t3_path, 'Projects')
+test_minimal_project_directory = os.path.join(PROJECTS_BASE_PATH, 'test_minimal_delete_after_usage')
```
Yes we use it in `test_main.py`, but since we only use it in `test_main.py`, I will define the path there instead.

The commit previously had moved that function, but one of your [PRs](https://github.com/ReactionMechanismGenerator/T3/commit/7a32790423e88e9dfd88822517f3c0bde11514a2) ended up doing that too so when I had rebased this branch, I had forgotten to update the commit message. I'll remove that phrase.

kspieks commented on t3/main.py:
```diff
@@ -119,6 +115,32 @@ class T3(object):
         rmg_species (List[Species]): Entries are RMG species objects in the model core for a certain T3 iteration.
         rmg_reactions (List[Reaction]): Entries are RMG reaction objects in the model core for a certain T3 iteration.
+        sa_dict (dict): Dictionary with keys of `kinetics`, `thermo`, and `time`. The full structure is given by:
+        sa_dict = { 'thermo' :
```
I moved it under `Dictionary structures::` and removed the line breaks.

kspieks commented on tests/test_simulate_adapters/data/rms_simulator_test/iteration_0/RMG/rms/chem9.rms:
```diff
@@ -0,0 +1,303 @@
+Phases:
```
No they're not necessary.

kspieks commented on t3/main.py:
```diff
@@ -265,11 +288,40 @@ def execute(self):
+                        sa_observables.append(species['label'])
+
+                # use simulate_factory to create the corresponding adapter for SA
+                if 'RMS' in self.t3['sensitivity']['adapter']:
```
I thought this would only work if each adapter was uniquely identified by the string used to register it?

alongd replied:
My understanding is that we could have adapters with the following names:
`RMG`
`RMS batch PT`
`RMS Flow HT`
`Cantera Batch liquid VT`
`Joes Cantera`

Since we should have different adapters for each type of reactor in RMS / Cantera, and we do allow for users to write their own, in which case the adapter name might not have a standard name.

If we have different visions for use cases, let's discuss it.

kspieks replied:
I see we can do that approach. To make sure I understand, you’re suggesting we have one adapter class (file) for each software and reactor type? Ex: `rms_constantV_simulator.py`, `rms_constantTP_simulator.py`, `cantera_constantTP_simulator.py`, `cantera_constantTP_simulator.py`, etc? And then register each of these with the factory was `RMS ConstantV`, `RMS ConstantTP`, etc?

In this case, the only difference between `RMS ConstantV` and `RMS ConstantTP` adapters etc will be one line where we define the domain. Same for `Cantera ConstantV` and `Cantera ConstantTP` etc.  So this will lead to duplicated code and tests that have to be maintained. But I agree that having one file per software and reactor type will make it easier for users to add their own, so I can reorganize the adapters in this way.

kspieks commented on t3/main.py:
```diff
@@ -265,11 +288,40 @@ def execute(self):
-                                      f"{self.t3['sensitivity']['adapter']}.")
-                    self.trsh_rmg_tol()
+                # determine species to run SA for
+                sa_observables = list()
```
Sure I'll update that

kspieks commented on t3/main.py:
```diff
@@ -265,11 +288,40 @@ def execute(self):
+
+                # use simulate_factory to create the corresponding adapter for SA
+                if 'RMS' in self.t3['sensitivity']['adapter']:
+                    simulate_method, rms_reactor_type = self.t3['sensitivity']['adapter'].strip().split(' ')
```
`simulate_method` simply identifies which software to use when simulating, not which reactor to use. The RMSSimulator class and CanteraSimulator class can each simulate different reactors, but each class is registered to the factory with only one string (unless we wanted to copy paste a bunch of code so each class only simulates one reactor type, but that seemed inefficient). So we need two strings. One to identify whether we use RMG, RMS, or Cantera (this is the `simulate_method`), and then another string to identify which specific reactor type that software uses. I'll update my typo in the factory docstring

alongd replied:
OK, this relates to my comment above. Seems like we have different points of view, yours does make sense, but on the other hand we should allow users to create their own adapters, I think. Let's talk about it to figure out how best to proceed.

kspieks commented on environment.yml:
```diff
@@ -8,6 +8,7 @@ channels:
   - omnia
   - pytorch
   - conda-forge
+  - mjohnson541
```
It's possible that Matt has already created an official version. But given the headache it took for us to get pyrms set up correctly, I'd prefer that we merge this branch as is and make any changes to pyrms on a future PR.

alongd replied:
OK. Could you post an issue and tag Matt as well, so we won't lose this?

kspieks replied:
Yes, just posted an issue to document our discussion from yesterday #67

kspieks commented on t3/common.py:
```diff
@@ -132,3 +132,27 @@ def time_lapse(t0: datetime.datetime) -> datetime.timedelta:
     return datetime.datetime.now() - t0
+
+
+def convert_termination_time_to_seconds(termination_time: Union[float, str]):
```
It should yes, thanks for catching this

kspieks commented on t3/simulate/rms_simulator.py:
```diff
@@ -0,0 +1,232 @@
+        return sa_dict
+
+
+    def get_idt(self):
```
I'll update the function name. I had added these just because it was easy to do with Cantera. T3 doesn't use these methods, so I hadn't added them to the factory.

kspieks commented on t3/simulate/rms_simulator.py:
```diff
@@ -0,0 +1,232 @@
+        self.phase = rms.IdealGas(self.spcs, self.rxns, name="gas")
+
+        # Define the domain (encodes how system thermodynamic properties calculated)
+        if self.rms_reactor_type == 'ConstantVDomain':
```
I'll add the `constantTVDomain` as well. I think that covers all of the basic domains

kspieks commented on t3/main.py:
```diff
@@ -739,63 +726,28 @@ def determine_species_based_on_sa(self) -> List[int]:
-                if sa_type is not None:
-                    # proceed only if we care about this column
+        # create new dictionary that stores the absolute maximum value from each SA
+        sa_dict_max = {'rxn': dict(), 'spc': dict()}
```
`sa_dict_max` is not an attribute of the T3 class since it is just an intermediate step to calculating `species_keys`. For this reason, I had thought it was okay to use this code since it's not used outside of this method

alongd replied:
right, it's not an attribute, but maybe we could still use consistent syntax?

kspieks replied:
Sure thing. I updated the keys to `kinetics` and `thermo` to be consistent with the keys in `sa_dict`

alongd reviewed (commented):
Thanks! I added some responses

kspieks commented on t3/simulate/rms_simulator.py:
```diff
@@ -0,0 +1,232 @@
+        self.set_up()
+
+
+    def set_up(self):
```
Thanks for noticing this. I'll add this functionality

kspieks commented on tests/test_main.py:
```diff
@@ -568,8 +584,31 @@ def test_determine_species_based_on_sa():
+                                        rtol=t3.rmg['model']['rtol'],
+                                        observable_list=sa_observables,
+                                        sa_atol=t3.t3['sensitivity']['atol'],
+                                        sa_rtol=t3.t3['sensitivity']['atol'],
```
Typo: this should say `rtol`. I will push this fix and address the unused import from LGTM with any additional feedback so that I only push once

kspieks commented on tests/test_simulate_adapters/test_rmg_simulator.py:
```diff
@@ -0,0 +1,115 @@
+
+    rmg_simulator_adapter = RMGSimulator(t3=t3.t3,
+                                        rmg=t3.rmg,
+                                        paths=t3.paths,
```
Minor: I will fix the indentation so everything aligns

kspieks commented on tests/test_simulate_adapters/test_rmg_simulator.py:
```diff
@@ -0,0 +1,115 @@
+                                        observable_list=observable_list,
+                                        sa_atol=t3.t3['sensitivity']['atol'],
+                                        sa_rtol=t3.t3['sensitivity']['atol'],
+                                        )
```
Minor: should say `rtol`

alongd commented on ipython/simulator_adapter_examples/cantera_simulator.ipynb:
```diff
@@ -0,0 +1,142 @@
+   "source": [
+    "# simulate ideal gas with constant T and P and perform sensitivity analysis wrt to rate constants and ethalpies\n",
+    "observable_list = ['OH', 'H']\n",
+    "CanteraSimulator_adapter = CanteraSimulatorConstantTP(t3=t3.t3,\n",
```
use lower_case_underscores for parameters?

kspieks replied:
Thanks for catching this. I changed all notebooks to use `cantera_simulator_adapter` for example.

alongd reviewed (commented):
Looks good! I added some minor comments.
Any ideas why the test are constantly "pending"? Hopefully this will sort out by pushing, later we'll make a transition.

kspieks commented on t3/simulate/factory.py:
```diff
@@ -0,0 +1,83 @@
+from typing import List, Optional, Type
+
+from t3.logger import Logger
+from t3.simulate.adapter import SimulateAdapter
```
Thanks for the tip :) Changing this now

kspieks replied:
I tried a few quick approaches, but it seems like `SimulateAdapter` must be imported for [line 28](https://github.com/ReactionMechanismGenerator/T3/pull/13/files#diff-7e6ef6a001df78c86d330e1583e0c2497a0a333880b171d77239d7f33aafa604R28) to run successfully. Let me know if this is okay.

kspieks commented on t3/simulate/factory.py:
```diff
@@ -0,0 +1,83 @@
+
+    if simulate_method not in _registered_simulate_adapters.keys():
+        raise ValueError(f'The "simulate_method" argument of {simulate_method} was not present in the keys for the '
+                         f'_registered_simulate_adapters dictionary: {_registered_simulate_adapters.keys()}'
```
Definitely, I'll update this

kspieks commented on tests/test_rmg_simulator.py:
```diff
@@ -0,0 +1,114 @@
+    dirs = [solver_directory, species_directory, sa_directory, log_archive]
+    for dir in dirs:
+        if os.path.isdir(dir):
+            shutil.rmtree(dir)
```
Absolutely :)

kspieks commented on t3/simulate/rms_constantP_simulator.py:
```diff
@@ -0,0 +1,217 @@
+"""
+RMS Simulator Adapter module
+Used to run mechanism analysis with RMS that simulates ideal gases at constant P
```
Thanks for catching this. Yes adiabatic and constant P should also mean constant enthalpy. I'll update the description

kspieks commented on t3/simulate/rms_constantV_simulator.py:
```diff
@@ -0,0 +1,217 @@
+"""
+RMS Simulator Adapter module
+Used to run mechanism analysis with RMS that simulates ideal gases at constant V
```
From talking with Matt, these are solved adiabatically since there's no heat transfer in or out of the domain. I think adiabatic is a more accurate description for constantV? I too want the description to be accurate and maybe I'm confusing myself

alongd replied:
Take a look at NASA CEA (a widely used software to predict products of fuels based on thermodynamics only). Look at "type code" in the from page. E.g., you have `hp` for Assigned Enthalpy & Pressure (so what we'll call `H-P`), and so on. We definitely don't have to conform to that, but it does reflect on the kinds of problems we care about. Constant V should go with constant U, so we can say "Constant V-U (adiabatic)".

mefuller replied:
I don't want to be obnoxious or pedantic, but we should conform to UV and HP (constant (internal) energy/constant volume & constant enthalpy/constant pressure) and while it's possible to pick a different pair of thermodynamic properties to hold constant, these two cases are drawn directly from the first law of thermodynamics and are essential standards.

kspieks commented on t3/simulate/cantera_constantP_simulator.py:
```diff
@@ -0,0 +1,540 @@
+"""
+Cantera Simulator Adapter module
+Used to run mechanism analysis with Cantera as ideal gases in a batch reactor at constant P
```
From looking on Cantera's website, constant enthalpy seems correct for the ideal gas constant P reactor. And yea, adiabatic and constant P should also mean constant enthalpy so thanks for pointing this out

kspieks commented on examples/commented/input.yml:
```diff
@@ -29,8 +29,8 @@ t3:
     global_observables: ['IgD', 'ESR', 'SL']  # optional, only implemented in the Cantera adapter, default: ``None``
-    SA_threshold: 0.01  # optional, default: 0.01
-    pdep_SA_threshold: 0.001  # optional, used to determine wells and reactions (to be implemented) to calculate
+    SA_threshold: 0.001  # optional, default: 0.001
```
I had thought in a previous conversation we discussed lowering the threshold. It's possible I misremembered. Should I undo this change?

alongd replied:
I might also misremember. I think it'll take (much) longer to run at a low tol. Can we justify it? If not, perhaps consult with Matt?

kspieks replied:
I'll change it back for now and will try asking Matt offline if it's worth changing the value later

kspieks commented:
Thanks for the review! I too am confused why the tests are pending...it was like this 4 days ago when I pushed the previous changes as well. It seems to be a recent change

alongd commented:
Thanks, @mefuller, for the comment. I agree that it makes sense (and correct) to conform to the UV / HP / TP syntax as much as possible, including the adapter names.

@kspieks, this PR looks great! Would you agree with Mark's comment? Also, did we modify the documentation? Would you please write a paragraph or two under [Writing simulation adapters](https://reactionmechanismgenerator.github.io/T3/how_to/#writing-simulation-adapters) to explain the logic to users and briefly say what needs to be done, and add a page to the [Tutorials](https://reactionmechanismgenerator.github.io/T3/tutorials/1_no_qm/)?

kspieks commented:
Yes I agree that standardizing the syntax makes sense so thanks for the suggestion. I renamed the classes and updated their corresponding docstrings. I also added documentation to those two sections

alongd commented:
I can ran the functional test fine locally using the same T3/RMG branches, strange that it began to fail here recently. I'm restarting the GH Actions tests.

alongd reviewed (commented):
@kspieks, could you please look at two minor comments?

kspieks commented on t3/simulate/factory.py:
```diff
@@ -0,0 +1,85 @@
+
+
+def register_simulate_adapter(simulator: str,
+                              simulate_class: Type['SimulateAdapter'],
```
Thanks I'll push this change if you agree with the change above :)

alongd commented:
@kspieks, everything looks good to me. Are we good to merge this in?

kspieks commented:
@alongd Yes I believe so. Thanks again for all of your helpful feedback in developing this PR. I'm happy to look at PR #69 after this is merged in as well as discuss other next steps

alongd reviewed (approved):
@kspieks, Thank you. This is a well deserved merge: a seminal PR which reflects a significant improvement in T3's abilities and flexibility. Thanks for the professional work! 🥇
